Make ocamlobjinfo work on .cmxs files on Windows

shindere opened the pull request:
For a long time, the BFD support was not enabled on Windows.

Recently, commit 69316099b041089c096816da2685b99c59d10f2e changed this
and so objinfo_helper started to be compiled on Windows and the related
test to be run... and to fail.

There are two reasons why the test was failing:

- objinfo_helper was looking for the wrong symbol, caml_plugin_header
  in stead of _caml_plugin_header

- It was looking for in in the dynamic symbol table while on Windows
  it can be found in the static symbol table

The PR addresses all this. The modification in `tools/Makefile` is to
look for the right symblo.

`objinfo_helper` has been modified to look in both symbol tables.

It should now be possible to use `objinfo` on `.cmxs` files on Windows.

shindere commented:
If accepted in trunk this should also be cherry-picked into 4.10.

dra27 commented:
Cool - I remember enabling the support for Cygwin, but never dug into why it wasn’t working! This is for the Cygwin port, not the native ports, though, right?

shindere commented:
David Allsopp (2019/12/12 11:56 -0800):
> Cool - I remember enabling the support for Cygwin, but never dug into
> why it wasn’t working! This is for the Cygwin port, not the native
> ports, though, right?

Well I don't know, to be honest.

I mean: it does work now on our cygwin32 slave, the one where the test
was failing before.

On cygwin64 bfd.h is found but not the library.

I didn't check the MinGW and MSVC ports yet but, at least in theory, I'd
say there should be a way to make it work too, no?

gasche reviewed (approved):
The code looks good to me. I don't understand the domain-specific stuff about "dynamic" versus "static" symbol stables, but the code is easy to read and does what it says on the tin, and this feature is covered by our testsuite so we know it does work better than before.

If think you should include a Changes entry for this PR, it is non-trivial work.

shindere commented:
Gabriel Scherer (2019/12/12 12:29 -0800):
> gasche approved this pull request.

Thanks!

> If think you should include a Changes entry for this PR, it is
> non-trivial work.

Done.

I'd be happy if somebody else could press the merge button and do the
cherry-picking. @Octachron, perhaps?

gasche commented:
The CI build fails on some configurations:

```
gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -fno-common -fno-tree-vrp -ffunction-sections -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -I../runtime '-Dsymbol_prefix=""'  -o objinfo_helper.o objinfo_helper.c
objinfo_helper.c: In function ‘error’:
objinfo_helper.c:73:3: error: implicit declaration of function ‘va_start’ [-Werror=implicit-function-declaration]
   va_start(ap, msg);
   ^
objinfo_helper.c:75:3: error: implicit declaration of function ‘va_end’ [-Werror=implicit-function-declaration]
   va_end(ap);
   ^
cc1: all warnings being treated as errors
../Makefile.common:78: recipe for target 'objinfo_helper.o' failed
```

Besides, `check-typo` also complains:

```
./tools/objinfo_helper.c:126.3: [white-at-eol] whitespace at end of line
```

shindere committed:
objinfo_helper: introduce an error function

shindere committed:
objinfo_hellper fix

This commit modifies objinfo_helper so that the caml_plugin_header
symbol is also looked up in the static symbol table, if it is
not found in the dynamic one.

This is useful e.g. on Windows where the symbol can not be found
without this commit.

It is also necessary to look for the _-prefixed version of the symbol.

shindere commented:
Gabriel Scherer (2019/12/14 04:33 -0800):
> The build fails on some configurations:

Oops, sorry about that! On which configurations did it fail, actually?

It should be fixed now.

> Besides, `check-typo` also complains:
>
> ```
> ./tools/objinfo_helper.c:126.3: [white-at-eol] whitespace at end of line
> ```

Fixed too, thanks a lot.

That's what you get when working on a computer where @dra27's excellent
pre-commit hook is not enabled.
